### PR TITLE
In progress: Improve idle logic

### PR DESCRIPTION
**Pull request by cramforce**

### CODE REVIEW

File: src/chunk.js [hunk 1]
```diff
@@ -202,6 +207,24 @@ class Chunks {
   }
 }
 
+function onIdle(win, startTime, minimumTimeRemaining, timeout, fn) {
+  win.requestIdleCallback(info => {
+    if (info.timeRemaining() < minimumTimeRemaining) {
+      const remainingTimeout = timeout - (Date.now() - startTime);
+      if (remainingTimeout <= 0) {
+        dev().fine('chunk', 'Timed out', timeout);
```
Comment (jridgewell) [inline src/chunk.js]: 'CHUNK'

Comment (cramforce) [inline src/chunk.js]: Done

**Comment by cramforce** [comment]
@jridgewell PTAL

File: src/chunk.js [hunk 2]
```diff
@@ -203,6 +216,39 @@ class Chunks {
...
+ * @param {number} timeout in millis for callback to fire.
+ * @param {function()} fn Callback.
+ * @visibleForTesting
+ */
+export function onIdle(win, startTime, minimumTimeRemaining, timeout, fn) {
+  win.requestIdleCallback(info => {
+    if (info.timeRemaining() < minimumTimeRemaining) {
+      const now = Date.now();
+      const remainingTimeout = timeout - (now - startTime);
+      if (remainingTimeout <= 0 || info.didTimeout) {
```
Comment (jridgewell) [inline src/chunk.js]: Won't these always be the same?

File: src/chunk.js [hunk 3]
```diff
@@ -203,6 +216,39 @@ class Chunks {
...
+        dev().fine(TAG, 'Timed out', timeout, info.didTimeout);
+        fn();
+      } else {
+        dev().fine(TAG, 'Rescheduling with', remainingTimeout,
+            info.timeRemaining());
+        onIdle(win, now, minimumTimeRemaining, remainingTimeout, fn);
```
Comment (jridgewell) [inline src/chunk.js]: We could make this less recursive by reusing the same function:

```js
function fn(info) {
  if (info.timeRemaining() < minimumTimeRemaining) {
    const now = Date.now();
    const remainingTimeout = timeout - (now - startTime);
    if (remainingTimeout <= 0) {
      fn();
    } else {
      requestIdleCallback(fn, { timeout: remainingTimeout });
    }
  } else {
    fn();
  }
}
requestIdleCallback(fn, { timeout: timeout });

Comment (cramforce) [inline src/chunk.js]: Done. Definitely nice to not need the `startTime` argument.

File: src/chunk.js (hunk 2 above)
Comment (cramforce) [inline src/chunk.js]: In theory yes.

**Comment by jridgewell** [comment]
LGTM.

**Pull request merged**